module/User/src/Service/ApiPrincipalService.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+declare(strict_types=1);
+
+namespace User\Service;
+
+use Laminas\Mvc\Plugin\FlashMessenger\FlashMessenger;
+use User\Form\ApiPrincipal as ApiPrincipalForm;
+use User\Mapper\ApiPrincipalMapper;
+use User\Model\ApiPrincipal as ApiPrincipalModel;
+use User\Model\Enums\ApiPermissions;
+
+use function array_map;
+
+class ApiPrincipalService
+{
+    public function __construct(
+        protected readonly ApiPrincipalMapper $mapper,
+        protected readonly ApiPrincipalForm $apiPrincipalForm,
+    ) {
+    }
+
+    public function getCreateForm(): ApiPrincipalForm
+    {
+        return $this->apiPrincipalForm;
+    }
+
+    public function getEditForm(ApiPrincipalModel $principal): ApiPrincipalForm
+    {
+        $this->apiPrincipalForm->bind($principal);
+
+        return $this->apiPrincipalForm;
+    }
+
+    /**
+     * @param array<array-key,mixed> $data
+     */
+    public function create(
+        array $data,
+        ?FlashMessenger $flashMessenger,
+    ): bool {
```

Comment:
Like I mentioned earlier, I would change this to return `ApiPrincipalModel|false` and remove the `$flashMessenger` parameter.

Reply:
Or `?ApiPrincipalModel` whatever makes the `if` in the controller cleaner (or check wat we do elsewhere).

Reply:
b48bbb2e9966b14482d8bd2b22ee043892724aac